### ISSUE TITLE
Music stored on samba sharing is not recognized

### ISSUE THREAD

**Issue by kronat** [gnumdk/lollypop#1033]
• Lollypop version: 0.9.230
• Gtk Version: 3.22.10
• Linux distribution: Archlinux

I have my music directory as a samba shared folder on my local network. I mount it with the following command:

`sudo mount.cifs //192.168.1.36/Music ~/Music -o user=nat`

and therefore I have my collection accessible to the user nat (read and write).

Now, if I run Lollypop (which I find wonderful, compared with my previous player, Banshee) only some of the files are recognized. In particular, I was thinking of some codec incompatibilies (I have mp3 as well as FLAC, Ogg, and some mac os types) but I noticed that the program isn't crawling inside the directories.

In particular, it shows the error "not detected as a music file" for each directory that I have in my Music folder, except some few. Here is an example, with an added debug print:

file:///home/nat/Music/ACDC detected as: enum G_FILE_TYPE_REGULAR of type Gio.FileType
file:///home/nat/Music/ACDC not detected as a music file

The ACDC folder is, obviously, a folder:

drwxr-xr-x+ 1 nat users 0 Feb 7 13:29 ACDC

which is reported as the same as the one that, instead, works (Bruce\ Springsteen/Human\ Touch):

drwxr-xr-x+ 1 nat users 0 Oct 31 2015 Bruce Springsteen
drwxr-xr-x+ 1 nat users 0 Oct 29 2015 Human Touch

I walked a little in the code and found that lollypop is using Lio class, which is a wrapper around Gio.File to detect the type. I tried to code a small python file that walks through my filesystem just using Gio, but the problem persists. Is probably better to report upstream?

Do you have any hints or suggestion that I can try?
[test.txt](https://github.com/gnumdk/lollypop/files/914054/test.txt)

Thanks!

(attached the test.py files)

**Comment by gnumdk** [comment]
In fact, there is an error in the code:
https://github.com/gnumdk/lollypop/commit/709facbe7d48d09c9e5d5d33164cc72bbc9e29de

Should allow us to understand what is happening ;)

**Comment by gnumdk** [comment]
I closed the issue as I'm quite sure last fix  76cd1b6 should have fixed the issue.

Please confirm, I will reopen if not the case.

**Comment by kronat** [comment]
Unfortunately, not. It has improved the situation (now the prints are good enough to understand what's going on). So, with the git version:

- if I put all my samba share to scan, it discards correctly jpg files; then, the audio content checking stops after the first (I suppose) directory. The directory is Ben\ Harper/Lifeline. If I add only such directory to the list, I have the following:
```
No module named 'pylast'
     - Scrobbler disabled
     - Auto cover download disabled
     - Artist informations disabled
 $ sudo pip3 install pylast
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_array_append_vals: assertion 'array' failed
  g_type = info.get_g_type()
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_hash_table_lookup: assertion 'hash_table != NULL' failed
  g_type = info.get_g_type()
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_hash_table_insert_internal: assertion 'hash_table != NULL' failed
  g_type = info.get_g_type()
Player::restore_state(): track missing
file:///home/nat/Music/Ben%20Harper/Lifeline/AlbumArtSmall.jpg not detected as a music file
file:///home/nat/Music/Ben%20Harper/Lifeline/Folder.jpg not detected as a music file
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/01.%20Fight%20Outta%20You.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/01.%20Fight%20Outta%20You.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/02.%20In%20the%20Colors.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/02.%20In%20the%20Colors.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/03.%20Fool%20for%20a%20Lonesome%20Train.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/03.%20Fool%20for%20a%20Lonesome%20Train.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/04.%20Needed%20You%20Tonight.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/04.%20Needed%20You%20Tonight.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/05.%20Having%20Wings.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/05.%20Having%20Wings.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/06.%20Say%20You%20Will.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/06.%20Say%20You%20Will.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/07.%20Younger%20Than%20Today.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/07.%20Younger%20Than%20Today.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/08.%20Put%20It%20on%20Me.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/08.%20Put%20It%20on%20Me.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/09.%20Heart%20of%20Matters.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/09.%20Heart%20of%20Matters.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/10.%20Paris%20Sunrise%20%237.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/10.%20Paris%20Sunrise%20%237.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Music/Ben%20Harper/Lifeline/11.%20Lifeline.flac' file:///home/nat/Music/Ben%20Harper/Lifeline/11.%20Lifeline.flac
```


I walked a little in the code, and I found the offending part:
```

diff --git i/src/collectionscanner.py w/src/collectionscanner.py
index 3fbde891..91cf225c 100644
--- i/src/collectionscanner.py
+++ w/src/collectionscanner.py
@@ -215,7 +215,7 @@ class CollectionScanner(GObject.GObject, TagReader):
                         if uri in orig_tracks:
                             orig_tracks.remove(uri)
                             i += 1
-                            if mtime <= mtimes[uri]:
+                            if uri in mtimes and mtime <= mtimes[uri]:
                                 i += 1
                                 continue
                             else:
```

With this patch applied, at least it tries to add the files:

```
No module named 'pylast'
    - Scrobbler disabled
    - Auto cover download disabled
    - Artist informations disabled
$ sudo pip3 install pylast
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_array_append_vals: assertion 'array' failed
  g_type = info.get_g_type()
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_hash_table_lookup: assertion 'hash_table != NULL' failed
  g_type = info.get_g_type()
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_hash_table_insert_internal: assertion 'hash_table != NULL' failed
  g_type = info.get_g_type()
Player::restore_state(): track missing
file:///home/nat/Music/Ben%20Harper/Lifeline/AlbumArtSmall.jpg not detected as a music file
file:///home/nat/Music/Ben%20Harper/Lifeline/Folder.jpg not detected as a music file
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/01.%20Fight%20Outta%20You.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/02.%20In%20the%20Colors.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/03.%20Fool%20for%20a%20Lonesome%20Train.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/04.%20Needed%20You%20Tonight.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/05.%20Having%20Wings.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/06.%20Say%20You%20Will.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/07.%20Younger%20Than%20Today.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/08.%20Put%20It%20on%20Me.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/09.%20Heart%20of%20Matters.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/10.%20Paris%20Sunrise%20%237.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
Adding file: file:///home/nat/Music/Ben%20Harper/Lifeline/11.%20Lifeline.flac
CollectionScanner::add2db(): Read tags
CollectionScanner::add2db(): Restore stats
CollectionScanner::add2db(): Add artists Ben Harper
CollectionScanner::add2db(): Add album artists 
CollectionScanner::add2db(): Add album: Lifeline, []
CollectionScanner::add2db(): Add track
CollectionScanner::add2db(): Update tracks
```

Unfortunately, I don't see anything in the UI. I think it is a problem that somewhere lies in the fact that the samba sharing is not exporting the same info as it would a normal FS. I asked a friend, he's using NFS and everything runs fine.

**Comment by gnumdk** [comment]
Strange because I already tested with Samba and here it works...

**Comment by gnumdk** [comment]
For me issue is in __get_objects_for_uris()
It set samba share as ignore dir so mtime is not available.

Can you check in get_objects_for_uris() what info.get_is_hidden() is returning?

**Comment by gnumdk** [comment]
#844

**Comment by gnumdk** [comment]
Can you send me your db, seems that issue is here, because if code go in if mtime <= mtimes[uri], track is already in db...

**Comment by gnumdk** [comment]
You send me an album but not tour ~/.local/share/lollypop/lollypop.db

**Comment by kronat** [comment]
[lollypop.db.zip](https://github.com/gnumdk/lollypop/files/944928/lollypop.db.zip)

Sorry for the late response. This is my current db; however, I don't see any of the songs in my UI. This is what I got as messages:

```
nat@judith ~ $ lollypop -d                                                                                                                                    

No module named 'pylast'
    - Scrobbler disabled
    - Auto cover download disabled
    - Artist informations disabled
$ sudo pip3 install pylast
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_array_append_vals: assertion 'array' failed
  g_type = info.get_g_type()
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_hash_table_lookup: assertion 'hash_table != NULL' failed
  g_type = info.get_g_type()
/usr/lib/python3.6/site-packages/gi/module.py:178: Warning: g_hash_table_insert_internal: assertion 'hash_table != NULL' failed
  g_type = info.get_g_type()
Player::restore_state(): track missing
file:///home/nat/Desktop/bellissima.jpg not detected as a music file
file:///home/nat/Lifeline/AlbumArtSmall.jpg not detected as a music file
file:///home/nat/Lifeline/Folder.jpg not detected as a music file
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/01.%20Fight%20Outta%20You.flac' file:///home/nat/Lifeline/01.%20Fight%20Outta%20You.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/02.%20In%20the%20Colors.flac' file:///home/nat/Lifeline/02.%20In%20the%20Colors.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/03.%20Fool%20for%20a%20Lonesome%20Train.flac' file:///home/nat/Lifeline/03.%20Fool%20for%20a%20Lonesome%20Train.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/04.%20Needed%20You%20Tonight.flac' file:///home/nat/Lifeline/04.%20Needed%20You%20Tonight.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/05.%20Having%20Wings.flac' file:///home/nat/Lifeline/05.%20Having%20Wings.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/06.%20Say%20You%20Will.flac' file:///home/nat/Lifeline/06.%20Say%20You%20Will.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/07.%20Younger%20Than%20Today.flac' file:///home/nat/Lifeline/07.%20Younger%20Than%20Today.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/08.%20Put%20It%20on%20Me.flac' file:///home/nat/Lifeline/08.%20Put%20It%20on%20Me.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/09.%20Heart%20of%20Matters.flac' file:///home/nat/Lifeline/09.%20Heart%20of%20Matters.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/10.%20Paris%20Sunrise%20%237.flac' file:///home/nat/Lifeline/10.%20Paris%20Sunrise%20%237.flac
CollectionScanner::__scan(mtime): 'file:///home/nat/Lifeline/11.%20Lifeline.flac' file:///home/nat/Lifeline/11.%20Lifeline.flac
```

**Comment by kronat** [comment]
After looking carefully, I discover that if I set a genre in the tracks (the .flac files I sent you did not contain any genre) I can see all the tracks in the UI.

I think the bug is not caused by the samba share, but in fact from the absence of the Genre in the tags. No genre, no mtime in the database, no display.

**Comment by gnumdk** [comment]
Will check soon

**Comment by gnumdk** [comment]
I'm sure it's not related to genre, but because you changed file mtime.

**Comment by gnumdk** [comment]
Can you send me by email your smb.conf on samba server.

**Comment by kronat** [comment]
I don't get the following: the mtime is saved in the genre table. If the song doesn't have a genre, how the mtime could be saved?

**Comment by gnumdk** [comment]
tagreader always set genre to "Unknown" if nothing available.

**Comment by kronat** [comment]
What happens to me is that, with the folder I sent you (regardless of the filesystem) at the first import (database empty) the "Unknown" category is not created.

**Comment by gnumdk** [comment]
regardless of the filesystem??? Local too?

**Comment by gnumdk** [comment]
If yes, can you send me folder again because it's not available anymore.

**Comment by gnumdk** [comment]
Thanks ;)

**Comment by kronat** [comment]
I've sent you the files; if you decompress the archive, import in an empty database just works. Now, close lollypop, delete the database and delete (or move to another location) the first song (with the name 01). Now, open the program and re-do the import from lollypop; no song will show up. Why? Because the first song was the only one with a Genre (Rock), which I added manually.

I just tried this on an ext4 filesystem, and this is the behavior I got.
